finagle/h2/src/main/scala/io/netty/handler/codec/http2/H2FrameStream.scala

```diff
@@ -0,0 +1,7 @@
+package io.netty.handler.codec.http2
+
+case class H2FrameStream(streamId: Int, streamState: Http2Stream.State) extends Http2FrameStream {
```

Comment:
Do you think that it would be worthwhile to encode the states a stream can be in as separate types rather than as a field? That way, we could encode valid state transitions at the type level. Consider something like this:
```scala
sealed abstract class H2FrameStream(override val state: Http2Stream.State) extends Http2FrameStream
object H2FrameStream {
  case class Open(id: Int) extends H2FrameStream(Http2Stream.State.OPEN)
  case class Closed(id: Int) extends H2FrameStream(Http2Stream.State.CLOSED)
  case class LocalClosed(id: Int) extends H2FrameStream(Http2Stream.State.HALF_CLOSED_LOCAL)
  case class RemoteClosed(id: Int) extends H2FrameStream(Http2Stream.State.HALF_CLOSED_REMOTE)
}

```
Then you can nicely pattern match on the variants without having to call `.state`, but you still get the `state` and `id` accessors that are required to implement `Http2FrameStream`.

Reply:
This sounds like a good idea, I am not familiar with how this would work but I can try it out and see what that is like.

Reply:
We don't encode any logic about these states and just pass them on to Netty so I don't think there's much to gain from more types here.  I think this probably can be made more concise...

```
case class H2FrameStream(streamId: Int, streamState: Http2Stream.State) extends Http2FrameStream
```

Since the case class will autogenerate vals for streamId and streamState that should satisfy the interface.